File: GXX.CSharp/src/GXX.LoginSrv/uFrmBatchEditAccountInfo.cs
using System;
using GXX.Core.Rtl;
using GXX.Core.Util;

namespace GXX.LoginSrv;

/// <summary>uFrmBatchEditAccountInfo.pas TNodeData（原文为 string[14]，赋值即截断）。</summary>
public sealed class TBatchEditNodeData
{
    private string _accountName = "";

    /// <summary>AccountName: string[14] —— 短字符串赋值截断（Delphi 语义）。</summary>
    public string AccountName
    {
        get => _accountName;
        set => _accountName = value != null && value.Length > 14 ? value.Substring(0, 14) : (value ?? "");
    }
}

/// <summary>
/// uFrmBatchEditAccountInfo.pas TFrmBatchEditAccountInfo 1:1（帐号批量启用/禁用）。
/// 源码：uFrmBatchEditAccountInfo.pas:56-327；DFM：uFrmBatchEditAccountInfo.dfm。
/// TVirtualStringTree → TVirtualStringTreeModel 接缝；TOpenDialog → OpenFileDialog + 注入提供者。
/// </summary>
public sealed class TFrmBatchEditAccountInfo : System.Windows.Forms.Form
{
    public System.Windows.Forms.Panel pnlLeft = null!;
    public TVirtualStringTreeModel vstEnableAccount = null!;
    public System.Windows.Forms.Panel Panel1 = null!;
    public System.Windows.Forms.Panel pnlRightTitle = null!;
    public System.Windows.Forms.Button btnSetEnabled = null!;
    public System.Windows.Forms.TextBox edtDisableAccount = null!;
    public System.Windows.Forms.Panel pnlLeftTitle = null!;
    public System.Windows.Forms.Button btnSetDisable = null!;
    public System.Windows.Forms.Button btn2 = null!;
    public System.Windows.Forms.TextBox edtEnabledAccount = null!;
    public System.Windows.Forms.Label lbl1 = null!;
    public System.Windows.Forms.Label Label1 = null!;
    public TVirtualStringTreeModel vstDisableAccount = null!;
    public System.Windows.Forms.OpenFileDialog dlgOpen = null!;

    /// <summary>测试注入：非 null 时替代 dlgOpen.Execute；返回 null 表示取消。</summary>
    public static Func<string?>? OpenFileProvider;

    public TFrmBatchEditAccountInfo()
    {
        InitializeComponent();
        FormCreate(this);
    }

    private void InitializeComponent()
    {
        // DFM: Caption = '批量编辑帐户信息'（uFrmBatchEditAccountInfo.dfm）
        Text = "批量编辑帐户信息";
        ClientSize = new System.Drawing.Size(640, 420);

        pnlLeft = new System.Windows.Forms.Panel { Left = 0, Top = 0, Width = 320, Height = 420, TabIndex = 0 };
        pnlLeftTitle = new System.Windows.Forms.Panel { Left = 0, Top = 0, Width = 320, Height = 24, TabIndex = 0 };
        lbl1 = new System.Windows.Forms.Label { Text = "可用帐户", Left = 6, Top = 5, Width = 80, Height = 12, AutoSize = false };
        pnlLeftTitle.Controls.Add(lbl1);
        edtEnabledAccount = new System.Windows.Forms.TextBox { Left = 90, Top = 2, Width = 150, Height = 20, TabIndex = 1 };
        edtEnabledAccount.TextChanged += (s, e) => edtEnabledAccountChange(s);
        pnlLeftTitle.Controls.Add(edtEnabledAccount);
        btnSetDisable = new System.Windows.Forms.Button { Text = "禁用选中(&D)", Left = 246, Top = 0, Width = 74, Height = 22, TabIndex = 2 };
        btnSetDisable.Click += (s, e) => btnSetDisableClick(s);
        pnlLeftTitle.Controls.Add(btnSetDisable);
        vstEnableAccount = new TVirtualStringTreeModel();
        btn2 = new System.Windows.Forms.Button { Text = "导入列表(&I)", Left = 0, Top = 0, Width = 90, Height = 22, TabIndex = 3 };
        btn2.Click += (s, e) => btn2Click(s);

        Panel1 = new System.Windows.Forms.Panel { Left = 320, Top = 0, Width = 320, Height = 420, TabIndex = 1 };
        pnlRightTitle = new System.Windows.Forms.Panel { Left = 0, Top = 0, Width = 320, Height = 24, TabIndex = 0 };
        Label1 = new System.Windows.Forms.Label { Text = "已禁用帐户", Left = 6, Top = 5, Width = 80, Height = 12, AutoSize = false };
        pnlRightTitle.Controls.Add(Label1);
        edtDisableAccount = new System.Windows.Forms.TextBox { Left = 90, Top = 2, Width = 150, Height = 20, TabIndex = 1 };
        edtDisableAccount.TextChanged += (s, e) => edtDisableAccountChange(s);
        pnlRightTitle.Controls.Add(edtDisableAccount);
        btnSetEnabled = new System.Windows.Forms.Button { Text = "解禁选中(&E)", Left = 246, Top = 0, Width = 74, Height = 22, TabIndex = 2 };
        btnSetEnabled.Click += (s, e) => btnSetEnabledClick(s);
        pnlRightTitle.Controls.Add(btnSetEnabled);
        vstDisableAccount = new TVirtualStringTreeModel();

        dlgOpen = new System.Windows.Forms.OpenFileDialog { Filter = "文本文件(*.txt)|*.txt", Title = "导入帐号列表" };

        Controls.Add(pnlLeft);
        Controls.Add(Panel1);
    }

    /// <summary>uFrmBatchEditAccountInfo.pas:56 ShowFrmBatchEditAccountInfo（全局过程）。</summary>
    public static void ShowFrmBatchEditAccountInfo(bool showModal = true)
    {
        var FrmBatchEditAccountInfo = new TFrmBatchEditAccountInfo();
        try
        {
            if (showModal) FrmBatchEditAccountInfo.ShowDialog();
        }
        finally
        {
            FrmBatchEditAccountInfo.Dispose();
        }
    }

    /// <summary>uFrmBatchEditAccountInfo.pas:68 FormCreate（g_AccountDB.GetAllAccount 分流两棵树）。</summary>
    public void FormCreate(object? Sender)
    {
        vstEnableAccount.NodeDataSize = 1;    // 原文 SizeOf(TNodeData)（string[14] 短串 = 15 字节）
        vstDisableAccount.NodeDataSize = 1;

        TStringList AccountList = new();
        if (LoginSrvShare.g_AccountDB != null)
            LoginSrvShare.g_AccountDB.GetAllAccount(AccountList);

        for (int I = 0; I <= AccountList.Count - 1; I++)
        {
            // 原文 Integer(AccountList.Objects[I]) = 0 → 启用列表，否则禁用列表
            if (Convert.ToInt32(AccountList.GetObject(I) ?? 0) == 0)
            {
                TVirtualNode Node = vstEnableAccount.AddChild(null);
                Node.CheckType = TCheckType.ctCheckBox;
                Node.CheckState = TCheckState.csUncheckedNormal;
                ((TBatchEditNodeData)EnsureData(Node)).AccountName = AccountList[I];
            }
            else
            {
                TVirtualNode Node = vstDisableAccount.AddChild(null);
                Node.CheckType = TCheckType.ctCheckBox;
                Node.CheckState = TCheckState.csUncheckedNormal;
                ((TBatchEditNodeData)EnsureData(Node)).AccountName = AccountList[I];
            }
        }
    }

    private static object EnsureData(TVirtualNode node) => node.Data ??= new TBatchEditNodeData();

    /// <summary>uFrmBatchEditAccountInfo.pas:108 vstEnableAccountGetText。</summary>
    public void vstEnableAccountGetText(TVirtualStringTreeModel Sender, TVirtualNode Node, int Column, ref string CellText)
    {
        TBatchEditNodeData? NodeData = Node?.Data as TBatchEditNodeData;
        if (NodeData == null) return;
        CellText = NodeData.AccountName;
    }

    /// <summary>uFrmBatchEditAccountInfo.pas:118 vstEnableAccountBeforeItemErase（奇数行斑马色 $00FFFBF7）。</summary>
    public void vstEnableAccountBeforeItemErase(TVirtualStringTreeModel Sender, TVirtualNode Node,
        ref int ItemColor, ref TItemEraseAction EraseAction)
    {
        if (Node.Index % 2 != 0)
        {
            ItemColor = 0x00FFFBF7;
            EraseAction = TItemEraseAction.eaColor;
        }
    }

    /// <summary>uFrmBatchEditAccountInfo.pas:130 btnSetDisableClick。</summary>
    public void btnSetDisableClick(object? Sender)
    {
        TStringList SL = new();
        TVirtualNode? Node = vstEnableAccount.GetFirstChecked();
        while (Node != null)
        {
            TBatchEditNodeData? NodeData = Node.Data as TBatchEditNodeData;
            SL.AddObject(NodeData?.AccountName ?? "", Node);
            Node = vstEnableAccount.GetNextChecked(Node);
        }

        if (SL.Count > 0)
        {
            if (LoginSrvShare.g_AccountDB != null && LoginSrvShare.g_AccountDB.EnabledAccounts(SL, false))
            {
                for (int I = 0; I <= SL.Count - 1; I++)
                {
                    TVirtualNode NewNode = vstDisableAccount.AddChild(null);
                    NewNode.CheckType = TCheckType.ctCheckBox;
                    NewNode.CheckState = TCheckState.csUncheckedNormal;
                    ((TBatchEditNodeData)EnsureData(NewNode)).AccountName = SL[I];

                    vstEnableAccount.DeleteNode(SL.GetObject(I) as TVirtualNode);
                }

                LoginSrvForms.ShowMessage("已成功禁用" + DelphiRTL.IntToStr(SL.Count) + "个帐户");
            }
        }
    }

    /// <summary>uFrmBatchEditAccountInfo.pas:171 btnSetEnabledClick。</summary>
    public void btnSetEnabledClick(object? Sender)
    {
        TStringList SL = new();
        TVirtualNode? Node = vstDisableAccount.GetFirstChecked();
        while (Node != null)
        {
            TBatchEditNodeData? NodeData = Node.Data as TBatchEditNodeData;
            SL.AddObject(NodeData?.AccountName ?? "", Node);
            Node = vstDisableAccount.GetNextChecked(Node);
        }

        if (SL.Count > 0)
        {
            if (LoginSrvShare.g_AccountDB != null && LoginSrvShare.g_AccountDB.EnabledAccounts(SL, true))
            {
                for (int I = 0; I <= SL.Count - 1; I++)
                {
                    TVirtualNode NewNode = vstEnableAccount.AddChild(null);
                    NewNode.CheckType = TCheckType.ctCheckBox;
                    NewNode.CheckState = TCheckState.csUncheckedNormal;
                    ((TBatchEditNodeData)EnsureData(NewNode)).AccountName = SL[I];

                    vstDisableAccount.DeleteNode(SL.GetObject(I) as TVirtualNode);
                }

                LoginSrvForms.ShowMessage("已成功解禁" + DelphiRTL.IntToStr(SL.Count) + "个帐户");
            }
        }
    }

    /// <summary>uFrmBatchEditAccountInfo.pas:212 edtEnabledAccountChange（空串→全可见）。</summary>
    public void edtEnabledAccountChange(object? Sender)
    {
        if (edtEnabledAccount.Text.Length == 0)
        {
            vstEnableAccount.BeginUpdate();
            try
            {
                TVirtualNode? Node = vstEnableAccount.GetFirst();
                while (Node != null)
                {
                    vstEnableAccount.SetVisible(Node, true);
                    Node = vstEnableAccount.GetNext(Node);
                }
            }
            finally
            {
                vstEnableAccount.EndUpdate();
            }
        }
        else
        {
            vstEnableAccount.BeginUpdate();
            try
            {
                TVirtualNode? Node = vstEnableAccount.GetFirst();
                while (Node != null)
                {
                    TBatchEditNodeData? NodeData = Node.Data as TBatchEditNodeData;
                    vstEnableAccount.SetVisible(Node, DelphiRTL.Pos(edtEnabledAccount.Text, NodeData?.AccountName ?? "") > 0);
                    Node = vstEnableAccount.GetNext(Node);
                }
            }
            finally
            {
                vstEnableAccount.EndUpdate();
            }
        }
    }

    /// <summary>uFrmBatchEditAccountInfo.pas:249 edtDisableAccountChange。</summary>
    public void edtDisableAccountChange(object? Sender)
    {
        if (edtDisableAccount.Text.Length == 0)
        {
            vstDisableAccount.BeginUpdate();
            try
            {
                TVirtualNode? Node = vstDisableAccount.GetFirst();
                while (Node != null)
                {
                    vstDisableAccount.SetVisible(Node, true);
                    Node = vstDisableAccount.GetNext(Node);
                }
            }
            finally
            {
                vstDisableAccount.EndUpdate();
            }
        }
        else
        {
            vstDisableAccount.BeginUpdate();
            try
            {
                TVirtualNode? Node = vstDisableAccount.GetFirst();
                while (Node != null)
                {
                    TBatchEditNodeData? NodeData = Node.Data as TBatchEditNodeData;
                    vstDisableAccount.SetVisible(Node, DelphiRTL.Pos(edtDisableAccount.Text, NodeData?.AccountName ?? "") > 0);
                    Node = vstDisableAccount.GetNext(Node);
                }
            }
            finally
            {
                vstDisableAccount.EndUpdate();
            }
        }
    }

    /// <summary>
    /// uFrmBatchEditAccountInfo.pas:286 btn2Click。
    /// ★ 两处原文怪癖保留：(a) 循环里写的是 `vstDisableAccount.IsVisible[Node]`，Node 却来自 vstEnableAccount；
    /// (b) `SL.IndexOf(...) > 0`（首行永不被选中，应为 &gt;= 0）。
    /// </summary>
    public void btn2Click(object? Sender)
    {
        string? FileName = OpenFileProvider != null
            ? OpenFileProvider()
            : (dlgOpen.ShowDialog() == System.Windows.Forms.DialogResult.OK ? dlgOpen.FileName : null);
        if (FileName != null)
        {
            TStringList SL = new();
            SL.LoadFromFile(FileName);

            int Count = 0;

            TVirtualNode? Node = vstEnableAccount.GetFirst();
            while (Node != null)
            {
                TBatchEditNodeData? NodeData = Node.Data as TBatchEditNodeData;

                vstDisableAccount.SetVisible(Node, true);   // 原文如此：写的是 vstDisableAccount

                if (SL.IndexOf(NodeData?.AccountName ?? "") > 0)
                {
                    Node.CheckState = TCheckState.csCheckedNormal;
                    Count++;
                }
                else
                {
                    Node.CheckState = TCheckState.csUnCheckedNormal;   // 原文如此（UnChecked 拼写）
                }

                Node = vstEnableAccount.GetNext(Node);
            }

            vstEnableAccount.Invalidate();
            LoginSrvForms.ShowMessage("已成功导入并选中" + DelphiRTL.IntToStr(Count) + "个帐户");
        }
    }
}

File: Source/LoginSrv/uFrmBatchEditAccountInfo.pas
unit uFrmBatchEditAccountInfo;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, ComCtrls, VirtualTrees, ExtCtrls, StdCtrls, JvgGroupBox, AccountDB;

type
  TFrmBatchEditAccountInfo = class(TForm)
    pnlLeft: TPanel;
    vstEnableAccount: TVirtualStringTree;
    Panel1: TPanel;
    pnlRightTitle: TPanel;
    btnSetEnabled: TButton;
    edtDisableAccount: TEdit;
    pnlLeftTitle: TPanel;
    btnSetDisable: TButton;
    btn2: TButton;
    edtEnabledAccount: TEdit;
    lbl1: TLabel;
    Label1: TLabel;
    vstDisableAccount: TVirtualStringTree;
    dlgOpen: TOpenDialog;
    procedure FormCreate(Sender: TObject);
    procedure vstEnableAccountGetText(Sender: TBaseVirtualTree;
      Node: PVirtualNode; Column: TColumnIndex; TextType: TVSTTextType;
      var CellText: WideString);
    procedure vstEnableAccountBeforeItemErase(Sender: TBaseVirtualTree;
      TargetCanvas: TCanvas; Node: PVirtualNode; ItemRect: TRect;
      var ItemColor: TColor; var EraseAction: TItemEraseAction);
    procedure btnSetDisableClick(Sender: TObject);
    procedure btnSetEnabledClick(Sender: TObject);
    procedure edtEnabledAccountChange(Sender: TObject);
    procedure edtDisableAccountChange(Sender: TObject);
    procedure btn2Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;


  procedure ShowFrmBatchEditAccountInfo;

implementation

{$R *.dfm}

type
  PNodeData = ^TNodeData;
  TNodeData = record
    AccountName: string[14];    // 帐户名
  end;

procedure ShowFrmBatchEditAccountInfo;
var
  FrmBatchEditAccountInfo: TFrmBatchEditAccountInfo;
begin
  FrmBatchEditAccountInfo := TFrmBatchEditAccountInfo.Create(nil);
  try
    FrmBatchEditAccountInfo.ShowModal;
  finally
    FrmBatchEditAccountInfo.Free;
  end;
end;

procedure TFrmBatchEditAccountInfo.FormCreate(Sender: TObject);
var
  I: Integer;
  AccountList: TStringList;

  Node: PVirtualNode;
  NodeData: PNodeData;
begin
  vstEnableAccount.NodeDataSize := SizeOf(TNodeData);
  vstDisableAccount.NodeDataSize := SizeOf(TNodeData);
  
  AccountList := TStringList.Create;
  try
    g_AccountDB.GetAllAccount(AccountList);

    for I := 0 to AccountList.Count - 1 do
    begin
      if Integer(AccountList.Objects[I]) = 0 then
      begin
        Node := vstEnableAccount.AddChild(nil);
        Node.CheckType := ctCheckBox;
        Node.CheckState := csUncheckedNormal;
        NodeData := vstEnableAccount.GetNodeData(Node);
        NodeData.AccountName := AccountList.Strings[I];
      end
      else
      begin
        Node := vstDisableAccount.AddChild(nil);
        Node.CheckType := ctCheckBox;
        Node.CheckState := csUncheckedNormal;
        NodeData := vstDisableAccount.GetNodeData(Node);
        NodeData.AccountName := AccountList.Strings[I];
      end;
    end;

  finally
    AccountList.Free;
  end;
end;

procedure TFrmBatchEditAccountInfo.vstEnableAccountGetText(
  Sender: TBaseVirtualTree; Node: PVirtualNode; Column: TColumnIndex;
  TextType: TVSTTextType; var CellText: WideString);
var
  NodeData: PNodeData;
begin
  NodeData := Sender.GetNodeData(Node);
  CellText := NodeData.AccountName;
end;

procedure TFrmBatchEditAccountInfo.vstEnableAccountBeforeItemErase(
  Sender: TBaseVirtualTree; TargetCanvas: TCanvas; Node: PVirtualNode;
  ItemRect: TRect; var ItemColor: TColor;
  var EraseAction: TItemEraseAction);
begin
  if Node.Index mod 2 <> 0 then
  begin
    ItemColor := $00FFFBF7;
    EraseAction := eaColor;
  end;
end;

procedure TFrmBatchEditAccountInfo.btnSetDisableClick(Sender: TObject);
var
  I: Integer;
  SL: TStringList;
  Node: PVirtualNode;
  NodeData: PNodeData;
begin
  SL := TStringList.Create;
  try
    Node := vstEnableAccount.GetFirstChecked();
    while Node <> nil do
    begin
      NodeData := vstEnableAccount.GetNodeData(Node);
      SL.AddObject(NodeData.AccountName, TObject(Node));
      Node := vstEnableAccount.GetNextChecked(Node);
    end;


    if SL.Count > 0 then
    begin
      if g_AccountDB.EnabledAccounts(SL, False) then
      begin
        for I := 0 to SL.Count - 1 do
        begin
          Node := vstDisableAccount.AddChild(nil);
          Node.CheckType := ctCheckBox;
          Node.CheckState := csUncheckedNormal;
          NodeData := vstDisableAccount.GetNodeData(Node);
          NodeData.AccountName := SL.Strings[I];

          vstEnableAccount.DeleteNode(PVirtualNode(SL.Objects[I]));
        end;

        ShowMessage('已成功禁用' + IntToStr(SL.Count) + '个帐户');
      end;
    end;
  finally
    SL.Free;
  end;
end;

procedure TFrmBatchEditAccountInfo.btnSetEnabledClick(Sender: TObject);
var
  I: Integer;
  SL: TStringList;
  Node: PVirtualNode;
  NodeData: PNodeData;
begin
  SL := TStringList.Create;
  try
    Node := vstDisableAccount.GetFirstChecked();
    while Node <> nil do
    begin
      NodeData := vstDisableAccount.GetNodeData(Node);
      SL.AddObject(NodeData.AccountName, TObject(Node));
      Node := vstDisableAccount.GetNextChecked(Node);
    end;


    if SL.Count > 0 then
    begin
      if g_AccountDB.EnabledAccounts(SL, True) then
      begin
        for I := 0 to SL.Count - 1 do
        begin
          Node := vstEnableAccount.AddChild(nil);
          Node.CheckType := ctCheckBox;
          Node.CheckState := csUncheckedNormal;
          NodeData := vstEnableAccount.GetNodeData(Node);
          NodeData.AccountName := SL.Strings[I];

          vstDisableAccount.DeleteNode(PVirtualNode(SL.Objects[I]));
        end;

        ShowMessage('已成功解禁' + IntToStr(SL.Count) + '个帐户');
      end;
    end;
  finally
    SL.Free;
  end;
end;

procedure TFrmBatchEditAccountInfo.edtEnabledAccountChange(
  Sender: TObject);
var
  Node: PVirtualNode;
  NodeData: PNodeData;
begin
  if Length(edtEnabledAccount.Text) = 0 then
  begin
    vstEnableAccount.BeginUpdate;
    try
      Node := vstEnableAccount.GetFirst();
      while Node <> nil do
      begin
        vstEnableAccount.IsVisible[Node] := True;
        Node := vstEnableAccount.GetNext(Node);
      end;
    finally
      vstEnableAccount.EndUpdate;
    end;
  end
  else
  begin
    vstEnableAccount.BeginUpdate;
    try
      Node := vstEnableAccount.GetFirst();
      while Node <> nil do
      begin
        NodeData := vstEnableAccount.GetNodeData(Node);
        vstEnableAccount.IsVisible[Node] := Pos(edtEnabledAccount.Text, NodeData.AccountName) > 0;
        Node := vstEnableAccount.GetNext(Node);
      end;
    finally
      vstEnableAccount.EndUpdate;
    end;
  end;
end;

procedure TFrmBatchEditAccountInfo.edtDisableAccountChange(
  Sender: TObject);
var
  Node: PVirtualNode;
  NodeData: PNodeData;
begin
  if Length(edtDisableAccount.Text) = 0 then
  begin
    vstDisableAccount.BeginUpdate;
    try
      Node := vstDisableAccount.GetFirst();
      while Node <> nil do
      begin
        vstDisableAccount.IsVisible[Node] := True;
        Node := vstDisableAccount.GetNext(Node);
      end;
    finally
      vstDisableAccount.EndUpdate;
    end;
  end
  else
  begin
    vstDisableAccount.BeginUpdate;
    try
      Node := vstDisableAccount.GetFirst();
      while Node <> nil do
      begin
        NodeData := vstDisableAccount.GetNodeData(Node);
        vstDisableAccount.IsVisible[Node] := Pos(edtDisableAccount.Text, NodeData.AccountName) > 0;
        Node := vstDisableAccount.GetNext(Node);
      end;
    finally
      vstDisableAccount.EndUpdate;
    end;
  end;
end;

procedure TFrmBatchEditAccountInfo.btn2Click(Sender: TObject);
var
  SL: TStringList;
  Node: PVirtualNode;
  NodeData: PNodeData;
  Count: Integer;
begin
  if dlgOpen.Execute then
  begin
    SL := TStringList.Create;
    try
      SL.LoadFromFile(dlgOpen.FileName);

      Count := 0;

      Node := vstEnableAccount.GetFirst();
      while Node <> nil do
      begin
        NodeData := vstEnableAccount.GetNodeData(Node);

        vstDisableAccount.IsVisible[Node] := True;

        if SL.IndexOf(NodeData.AccountName) > 0 then
        begin
          Node.CheckState := csCheckedNormal;
          Inc(Count);
        end
        else
        begin
          Node.CheckState := csUnCheckedNormal;
        end;

        Node := vstEnableAccount.GetNext(Node);
      end;

      vstEnableAccount.Invalidate;
      ShowMessage('已成功导入并选中' + IntToStr(Count) + '个帐户');
    finally
      SL.Free;
    end;
  end;
end;

end.
